notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,347 @@
+types and returns `Just (b, args)` where `b` is the name of the
+builtin and `args` is a list of the arguments (and returns `Nothing`
+if is is not of this form, leading to case 3 above).  In the original
+CEK machine, this information was then passed to the CAM, which
+retrieves the Haskell code for the appropriate builtin and applies it
+the arguments, leading to one of three responses: `Stuck` if there
+were too few arguments, an error if there were too many arguments, and
+a `Success` result containing the result of applying the builtin
+if there are the correct number of arguments and they have the correct
+types.
+
+This process is quite inefficient: it usually requires repeated
+applications of `termAsPrimIterApp` and repeated calls to the CAM.
+
+### Saturated builtins
+
+One way to solve this problem is to insist on saturated builtins: the
+typechecker can check that all builtin applications have the correct
+number of arguments, and the evaluator is provided with an AST node
+containing the correct number of arguments, so `termAsPrimIterApp`
+isn't required and only a single call to the CAM is required.  On
+simple benchmarks involving a lot of builtin applications, this was
+substantially faster than the original CEK machine.  However,
+saturated builtins complicated matters considerably and we decided not
+to pursue them: see [this document](../../../language-plutus-core/docs/Saturatedness.md)
+for a consideration of the issues involved.
+One important factor was that an alternative version of the CEK
+machine allowed us to retain unsaturated builtins but achieve similar
+performance to unsaturated builtins.
+
+### The alternative CEK machine Part of the reason for abandoning
+saturated builtins was that James proposed an alternative version of
+the CEK machine which looked as if it might be more efficient than the
+original machine with respect to environment handling. This involved a
+new notion of "value" which also allowed us to simplify builtin
+application by giving us somewhere to store the arguments to which a
+builtin had so far been applied (this is an issue orthogonal to
+handling environments: we could have retained the original method
+of builtin evaluation and used a simpler version of the `builtin`
+value below while still improving the handling of environments).
+
+The original machine dealt with closures: pairs `(V,ρ)` with `V` a value and `ρ`
+an environment containing values for the free variables in `V` (and probably
+other variables as well: the new machine appeared attractive because it would
+reduce the number of irrelevant variables stored in environments).
+
+In the new proposal, closures are replaced with a new kind of "value":
+
+```
+V := con cn
+   | abs ɑ (M , ⍴)
+   | lam x (M , ρ)
+   | iwrap A B V
+   | builtin ρ bn count [types] [V]
+```
+(we'll call these things "CEK-values")
+
+These are similar to, but distinct from, the "values" in the Plutus Core
```

Comment:
I cannot resist nit-picking about the values not reducing further. I think that both version could reduce further if we reduced under binders. I think it's easier to point out that the ordinary "values" have just terms under binders.